### ISSUE TITLE
Refactoring snapshots

### ISSUE THREAD

**Issue by boriskovar-m2ms** [m2ms/fragalysis-frontend#1483]

**Comment by Waztom** [comment]
@boriskovar-m2ms has started to review the time/steps needed for this refcatoring excercise and estimates update next Tuesday.

**Comment by boriskovar-m2ms** [comment]
[SnapshotsRework-Estimations.xlsx](https://github.com/user-attachments/files/16597259/SnapshotsRework-Estimations.xlsx)
My estimations of refactoring.

**Comment by mwinokan** [comment]
Thanks for the estimate @boriskovar-m2ms.

@Waztom and I discussed the refactor and we think you should go ahead on this.

**Comment by mwinokan** [comment]
@boriskovar-m2ms has been thinking about the approach to the state-based snapshot resolution, and will begin work imminently.

An update on the time estimate: 18 working days would resolve to 6 weeks as Boris is working 3 days per week on Fragalysis.

**Comment by mwinokan** [comment]
@boriskovar-m2ms has completed the state-based rendering of ligands, and is working ligand removal. The rest of the rendering elements are pending. This would round off the NGL changes necessary for the rest of the refactor.

**Comment by mwinokan** [comment]
@boriskovar-m2ms hit a roadblock with rendering densities, that only occurs when rendering locally, but it has been fixed. Once this is done, Boris will move on to the RHS, after which the groundwork for the snapshot refactor itself will complete.

**Comment by mwinokan** [comment]
Last week @boriskovar-m2ms fixed the electron density and atom quality rendering, and will work on rendering ligands, complexes, surfaces and interactions.

**Comment by mwinokan** [comment]
@boriskovar-m2ms confirms that state based rendering is complete for the RHS. Now Boris can workshop sending state to the backend

**Comment by mwinokan** [comment]
@boriskovar-m2ms says he can upload the state to the b/e but there is outstanding work (being skipped for now) regarding the copying the state (have to work around hard limit of 2GB RAM usage e.g. in Chrome), Boris wants to have the end-to-end functionality first and then will craft a copy algorithm that works within the memory limits.

**Comment by mwinokan** [comment]
@boriskovar-m2ms reports that restoring orientations has been more challenging. Some of the effects for opening modals happen in a random order, and some effects require data to be loaded which is not generally true. Currently, Boris is dealing with a too large request header error - hopefully a fresh Docker install will resolve the discrepancy between working Tibor stack and Boris' local build (especially as the Tibor stack header is larger than the local one).

**Comment by mwinokan** [comment]
@matej-vavrek reports that @boriskovar-m2ms is almost done, working on restoring scroll functionality, and switching between snapshots. Optimistically, could be all done next week

@phraenquex asks that the NGL view should smoothly interpolate between the two orientations when switching between snapshots

**Comment by mwinokan** [comment]
@boriskovar-m2ms says that scrolling is working, and he is working on restoring the saving the project functionality (which is very similar to saving snapshots). Boris will also work on restoring the interpolation between snapshots in the projects.

**Comment by mwinokan** [comment]
@boriskovar-m2ms is working on saving and restoring projects

**Comment by mwinokan** [comment]
@boriskovar-m2ms is unwell but he will finish switching between the snapshots and then the refactor is complete

**Comment by mwinokan** [comment]
@boriskovar-m2ms is back to working on changing between snapshots, after which we will be feature complete and will implement the [optimisation](https://github.com/m2ms/fragalysis-frontend/issues/1483#issuecomment-2355375173)

**Comment by mwinokan** [comment]
@boriskovar-m2ms has had to rebase with the latest b/e and f/e images and is dealing with many conflicts. 

@boriskovar-m2ms asks if it is possible to upload data (LHS & RHS) to a local stack if ISpyB is down? @alanbchristie says an environment variable might be needed to assign Boris to a specific proposal (without ISpyB) and Alan can help Boris achieve this.

**Comment by mwinokan** [comment]
@boriskovar-m2ms has completed merge/rebase and resolved conflicts (lots of phantom errors). Related to using Frank's SSH connection (where there seemed to be some throttling/timeouts) caused some false negatives on f/e bugs.

Work is still in progress to make the snapshot transitions smooth(er). **Aims to finish it this week**, but there are more difficulties than expected, even after the refactor. The transition actually happens between three states, current, target, and a "phantom" state that preserves some, but not all, of the current state.

**Comment by Waztom** [comment]
Boris adding save functionality after changes made to the snapshot. Frank suggests continuously capturing changes with "snapshot saved" toast notification (edit or delete) to then link to snapshot editing modal vs. option to save modal.

**Comment by mwinokan** [comment]
@boriskovar-m2ms has completed almost everything, the only outstanding work is that changes to an existing snapshot should be saved when working within a project (about 50% complete). 

Meanwhile, Boris has also been squashing bugs (some of which are a consequence of the memory optimisation). One bug remains open: when you are logged in and share a snapshot to someone who is also logged in who then tries to open it, it might fail subtly or spectacularly (couple hours of work pending).

**Comment by Waztom** [comment]
@boriskovar-m2ms deployment is stuck...pending is one final functionality, needs to add a button/link to the snapshot save/edit name modal.

**Comment by boriskovar-m2ms** [comment]
deployed to tibor's stack

**Comment by mwinokan** [comment]
@phraenquex reports that the NGL rotation is not restored in Firefox

@boriskovar-m2ms says that he has a very large backlog of work on other projects which he will work on exclusively next week, but we should enumerate a shopping list of bugs to squash for when he's back

**Comment by phraenquex** [comment]
Bugs spotted by @phraenquex: 

- [ ] When I create a New Project, don't redraw everything - just make it be a project, and leave the UI as is.
- [ ]

**Comment by mwinokan** [comment]
@boriskovar-m2ms is working on @phraenquex's [bugs](https://github.com/m2ms/fragalysis-frontend/issues/1483#issuecomment-2470400078)

**Comment by mwinokan** [comment]
@boriskovar-m2ms is finalising the work today, and will create a PR to merge to staging

**Comment by boriskovar-m2ms** [comment]
All known bugs are resolved now and PR is waiting here https://github.com/xchem/fragalysis-frontend/pull/461/ which can be merged after stuff from staging is merged to production

**Comment by mwinokan** [comment]
Merged to staging

**Comment by mwinokan** [comment]
Unfortunately I managed to create some broken snapshots when RHS compound sets don't contain any molecules: https://fragalysis.xchem.diamond.ac.uk/viewer/react/projects/198/191

But this seems like a non-issue as those sets should be deleted anyway

**Comment by boriskovar-m2ms** [comment]
First one is crashing because there are datasets on RHS with no molecules? Should I fix it? Because I think such datasets shouldn't even be served to the frontend because they're clearly broken.

Second one works for me (tried like 10 times).

Third one - I can confirm that density is not rendered.

**Comment by mwinokan** [comment]
Thanks for confirming @boriskovar-m2ms 

Weirdly all three of them crashed occasionally for me but I'm just going to ignore that fact as reloading seemed to help. 

No need to support empty compound sets

**Comment by boriskovar-m2ms** [comment]
@mwinokan I know what's the problem with densities and why they worked for me before. I think now I know how to fix it but I don't think that's something I'll finish today.

**Comment by mwinokan** [comment]
@boriskovar-m2ms No problem. I'll make another ticket

**Comment by boriskovar-m2ms** [comment]
The issue is that for our convenience we at some places extending data downloaded from backend with some operational data and when saving snapshot we are striping the state of all downloaded data so this operational data will get lost so know I need to move this information somewhere else and with it also chunk of the code.

But I think there might be more of this kind of stinkers (maybe two or three other cases and I'm trying to remember).

**Comment by mwinokan** [comment]
See spinout ticket #1603
